### ISSUE TITLE
Flow.zip() - issue with kotlin.Result

### ISSUE THREAD

**Issue by zlnaga** [Kotlin/kotlinx.coroutines#2268]
Version: 
kotlin: `1.4.10`
coroutine: `1.3.9`

There are two flows:
```kotlin
val flow1 = flowOf(1, 2, 3, 4, 5)
val flow2 = flowOf("a", "b", "c", "d", "e")
```
when I try:
```kotlin
runBlocking {
  flow1.zip(flow2) { num, le -> println("$le$num") }.collect()
}
```

it output normally as:
```shell
a1
b2
c3
d4
e5
```

if I use map to transform by calling `Result.success` the data, like:
```kotlin
flow1.map { Result.success(it) }.zip(flow2) { numResult, le ->
  println("$le${numResult.getOrNull() ?: "?"}")
}.collect()
```

the output face an error: 
```shell
Caused by: java.lang.ClassCastException: kotlin.Result cannot be cast to java.lang.Integer
	at xxx.ExampleUnitTest$test_zip$1$2.invokeSuspend(ExampleUnitTest.kt:121)
	at xxx.ExampleUnitTest$test_zip$1$2.invoke(ExampleUnitTest.kt)
	at kotlinx.coroutines.flow.internal.CombineKt$zipImpl$$inlined$unsafeFlow$1$lambda$1.invokeSuspend(Combine.kt:148)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at xxx.ExampleUnitTest.test_zip(ExampleUnitTest.kt:113)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
```

but if I transform to a data class, like:
```kotlin
data class MyResult<T>(
  private val data: T?
) {
  fun getOrNull(): T? = data
}

flow1.map { MyResult(it) }.zip(flow2) { numResult, le ->
  println("$le${numResult.getOrNull() ?: "?"}")
}.collect()
```
everything works fine

Any idea?

**Comment by qwwdfsad** [comment]
Unfortunately, It is a compiler bug https://youtrack.jetbrains.com/issue/KT-41429 related to inline-classes.
It will be fixed in Kotlin 1.4.20 and Kotlin 1.4.20-M1.

**Comment by zlnaga** [comment]
> Unfortunately, It is a compiler bug https://youtrack.jetbrains.com/issue/KT-41429 related to inline-classes.
> It will be fixed in Kotlin 1.4.20 and Kotlin 1.4.20-M1.

Thanks for the reply and the link ref, will follow the ticket
